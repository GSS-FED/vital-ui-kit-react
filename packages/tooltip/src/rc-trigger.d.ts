declare module 'rc-trigger' {
  import React from 'react';
  export interface RcTriggerProps {
    /** whether popup is visible */
    popupVisible?: boolean;
    /** which actions cause popup shown */
    action?: Array<'hover' | 'click' | 'focus' | 'contextMenu'>;
    popupAlign?: PopupAlign;
    placement?: Placement;
    /** Popup content */
    popup: React.ReactNode | (() => React.ReactNode);
    /** additional style of popup */
    popupStyle?: React.CSSProperties;
    /** additional className added to popup */
    popupClassName?: string;
    /** builtin placement align map. used by placement prop */
    builtinPlacements?: object;
    zIndex?: number;
    /** https://github.com/react-component/animate */
    popupTransitionName?: string | object;
    maskTransitionName?: string | object;
    /** delay time to show when mouse enter. unit: s. */
    mouseEnterDelay?: number;
    /** delay time to hide when mouse leave. unit: s. */
    mouseLeaveDelay?: number;
    /** whether destroy popup when hide */
    destroyPopupOnHide?: boolean;
    /** 'width', 'minWidth', 'height', 'minHeight' */
    stretch?: string;
    /** use preset popup align config from builtinPlacements, can be merged by popupAlign prop */
    popupPlacement?: any;
  }

  export default class extends React.Component<Props> {}
}
