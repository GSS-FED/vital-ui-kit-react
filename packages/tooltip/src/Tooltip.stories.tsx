import React from 'react';
import { storiesOf } from '@storybook/react';
import { Tooltip } from '.';
import 'rc-trigger/assets/index.css';

storiesOf('Packages | Tooltip', module).add('Basic', () => (
  <Tooltip popup={<div>hi</div>}>
    <button>hover</button>
  </Tooltip>
));
