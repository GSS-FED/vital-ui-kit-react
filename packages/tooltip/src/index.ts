export * from './Tooltip';
export * from './styled';
