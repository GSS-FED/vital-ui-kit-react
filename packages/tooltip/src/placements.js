const autoAdjustOverflow = {
  adjustX: 1,
  adjustY: 1,
};

const targetOffset = [0, 0];

// eslint-disable-next-line
export const placements = {
  left: {
    points: ['cr', 'cl'],
    overflow: autoAdjustOverflow,
    offset: [-10, 0],
    targetOffset,
  },
  right: {
    points: ['cl', 'cr'],
    overflow: autoAdjustOverflow,
    offset: [10, 0],
    targetOffset,
  },
  top: {
    points: ['bc', 'tc'],
    overflow: autoAdjustOverflow,
    offset: [0, -10],
    targetOffset,
  },
  bottom: {
    points: ['tc', 'bc'],
    overflow: autoAdjustOverflow,
    offset: [0, 10],
    targetOffset,
  },
  topLeft: {
    points: ['bl', 'tl'],
    overflow: autoAdjustOverflow,
    offset: [0, -10],
    targetOffset,
  },
  leftTop: {
    points: ['tr', 'tl'],
    overflow: autoAdjustOverflow,
    offset: [-10, 0],
    targetOffset,
  },
  topRight: {
    points: ['br', 'tr'],
    overflow: autoAdjustOverflow,
    offset: [0, -10],
    targetOffset,
  },
  rightTop: {
    points: ['tl', 'tr'],
    overflow: autoAdjustOverflow,
    offset: [10, 0],
    targetOffset,
  },
  bottomRight: {
    points: ['tr', 'br'],
    overflow: autoAdjustOverflow,
    offset: [0, 10],
    targetOffset,
  },
  rightBottom: {
    points: ['bl', 'br'],
    overflow: autoAdjustOverflow,
    offset: [10, 0],
    targetOffset,
  },
  bottomLeft: {
    points: ['tl', 'bl'],
    overflow: autoAdjustOverflow,
    offset: [0, 10],
    targetOffset,
  },
  leftBottom: {
    points: ['br', 'bl'],
    overflow: autoAdjustOverflow,
    offset: [-10, 0],
    targetOffset,
  },
};
