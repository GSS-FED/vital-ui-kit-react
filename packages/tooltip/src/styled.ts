import { css } from 'styled-components';
import { defaultTheme } from '@vital-ui/react-theme';

export const rcTooltipStyle = ({
  theme = defaultTheme,
}: {
  theme?: typeof defaultTheme;
}) => css`
  .rc-tooltip-arrow {
    position: absolute;
    width: 0;
    height: 0;
    border-color: transparent;
    border-style: solid;
  }
  .rc-trigger-popup-placement-top .rc-tooltip-arrow,
  .rc-trigger-popup-placement-topLeft .rc-tooltip-arrow,
  .rc-trigger-popup-placement-topRight .rc-tooltip-arrow {
    bottom: -5px;
    margin-left: -5px;
    border-width: 5px 5px 0;
    border-top-color: ${theme.tooltip.bg};
  }
  .rc-trigger-popup-placement-top .rc-tooltip-arrow {
    left: 50%;
  }
  .rc-trigger-popup-placement-topLeft .rc-tooltip-arrow {
    left: 15%;
  }
  .rc-trigger-popup-placement-topRight .rc-tooltip-arrow {
    right: 15%;
  }
  .rc-trigger-popup-placement-right .rc-tooltip-arrow,
  .rc-trigger-popup-placement-rightTop .rc-tooltip-arrow,
  .rc-trigger-popup-placement-rightBottom .rc-tooltip-arrow {
    left: -5px;
    margin-top: -5px;
    border-width: 5px 5px 5px 0;
    border-right-color: ${theme.tooltip.bg};
  }
  .rc-trigger-popup-placement-right .rc-tooltip-arrow {
    top: 50%;
  }
  .rc-trigger-popup-placement-rightTop .rc-tooltip-arrow {
    top: 15%;
    margin-top: 0;
  }
  .rc-trigger-popup-placement-rightBottom .rc-tooltip-arrow {
    bottom: 15%;
  }
  .rc-trigger-popup-placement-left .rc-tooltip-arrow,
  .rc-trigger-popup-placement-leftTop .rc-tooltip-arrow,
  .rc-trigger-popup-placement-leftBottom .rc-tooltip-arrow {
    right: -5px;
    margin-top: -5px;
    border-width: 5px 0 5px 5px;
    border-left-color: ${theme.tooltip.bg};
  }
  .rc-trigger-popup-placement-left .rc-tooltip-arrow {
    top: 50%;
  }
  .rc-trigger-popup-placement-leftTop .rc-tooltip-arrow {
    top: 15%;
    margin-top: 0;
  }
  .rc-trigger-popup-placement-leftBottom .rc-tooltip-arrow {
    bottom: 15%;
  }
  .rc-trigger-popup-placement-bottom .rc-tooltip-arrow,
  .rc-trigger-popup-placement-bottomLeft .rc-tooltip-arrow,
  .rc-trigger-popup-placement-bottomRight .rc-tooltip-arrow {
    top: -5px;
    margin-left: -5px;
    border-width: 0 5px 5px;
    border-bottom-color: ${theme.tooltip.bg};
  }
  .rc-trigger-popup-placement-bottom .rc-tooltip-arrow {
    left: 50%;
  }
  .rc-trigger-popup-placement-bottomLeft .rc-tooltip-arrow {
    left: 15%;
  }
  .rc-trigger-popup-placement-bottomRight .rc-tooltip-arrow {
    right: 15%;
  }
`;
