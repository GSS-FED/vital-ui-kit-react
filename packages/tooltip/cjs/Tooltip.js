'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _templateObject = _taggedTemplateLiteral(['\n  display: inline-block;\n  padding: 8px 16px;\n  text-align: center;\n  color: ', ';\n  background: ', ';\n  border-radius: 4px;\n'], ['\n  display: inline-block;\n  padding: 8px 16px;\n  text-align: center;\n  color: ', ';\n  background: ', ';\n  border-radius: 4px;\n']),
    _templateObject2 = _taggedTemplateLiteral([''], ['']);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _rcTrigger = require('rc-trigger');

var _rcTrigger2 = _interopRequireDefault(_rcTrigger);

require('rc-trigger/assets/index.css');

require('./tooltip.css');

var _placements = require('./placements');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); } /**
                                                                                                                                                   * 
                                                                                                                                                   * Copyright Â© 2018 Galaxy Software Services https://github.com/GSS-FED/vital-ui-kit-react
                                                                                                                                                   * MIT license
                                                                                                                                                   */

var TooltipContent = _styledComponents2.default.div(_templateObject, function (_ref) {
  var theme = _ref.theme;
  return theme.tooltip.color;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.tooltip.bg;
});

var TooltipArrow = _styledComponents2.default.div(_templateObject2);

var Tooltip = function Tooltip(_ref3) {
  var children = _ref3.children,
      trigger = _ref3.trigger,
      overlay = _ref3.overlay,
      mouseEnterDelay = _ref3.mouseEnterDelay,
      mouseLeaveDelay = _ref3.mouseLeaveDelay,
      popupStyle = _ref3.popupStyle,
      popupTransitionName = _ref3.popupTransitionName,
      zIndex = _ref3.zIndex,
      popupAlign = _ref3.popupAlign,
      popupClassName = _ref3.popupClassName,
      destroyPopupOnHide = _ref3.destroyPopupOnHide,
      placement = _ref3.placement;

  var TooltipBase = function TooltipBase() {
    return [React.createElement(TooltipArrow, { className: 'rc-tooltip-arrow', key: 'arrow' }), React.createElement(
      TooltipContent,
      _extends({ key: 'content' }, popupStyle),
      overlay
    )];
  };

  return React.createElement(
    _rcTrigger2.default,
    {
      popup: TooltipBase,
      action: trigger,
      mouseEnterDelay: mouseEnterDelay,
      mouseLeaveDelay: mouseLeaveDelay,
      popupStyle: popupStyle,
      popupTransitionName: popupTransitionName,
      zIndex: zIndex,
      popupAlign: popupAlign,
      popupClassName: popupClassName,
      destroyPopupOnHide: destroyPopupOnHide,
      popupPlacement: placement,
      builtinPlacements: _placements.placements
    },
    children
  );
};

Tooltip.defaultProps = {
  overlay: null,
  mouseEnterDelay: 0,
  mouseLeaveDelay: 0,
  popupStyle: {},
  popupTransitionName: '',
  zIndex: null,
  popupAlign: null,
  popupClassName: 'vital-popup',
  destroyPopupOnHide: false
};

exports.default = Tooltip;