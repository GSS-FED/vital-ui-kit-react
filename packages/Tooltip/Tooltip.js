/**
 * @flow
 * Copyright Â© 2017 Galaxy Software Services https://github.com/GSS-FED/vital-ui-kit-react
 * MIT license
 */

import * as React from 'react';
import styled from 'styled-components';
import Trigger from 'rc-trigger';
import 'rc-trigger/assets/index.css';
import './tooltip.css';

import { placements } from './placements';

const TooltipContent = styled.div`
  display: inline-block;
  padding: 8px 16px;
  text-align: center;
  color: #fff;
  background: #182d54;
  border-radius: 4px;
`;

const TooltipArrow = styled.div``;

type Placement =
  | 'left'
  | 'right'
  | 'top'
  | 'bottom'
  | 'topLeft'
  | 'leftTop'
  | 'topRight'
  | 'rightTop'
  | 'bottomRight'
  | 'rightBottom'
  | 'bottomLeft'
  | 'leftBottom';

type Props = {
  /** the element target for tooltip */
  children: React.Node,
  /** an array of 'click', 'hover', 'focus', 'contextMenu' */
  trigger: Array<string>,
  /** the placement tooltip show up */
  placement: Placement,
  /** popup overlay content, with default styled wrapper */
  overlay?: React.Node,
  /** render your own custom overlay with no default style */
  mouseEnterDelay?: number,
  mouseLeaveDelay?: number,
  /** insert popup style */
  popupStyle?: Object,
  popupTransitionName?: String | Object,
  /** popup z-index */
  zIndex?: number,
  popupAlign?: Object,
  popupClassName?: string,
  destroyPopupOnHide?: boolean
};

const Tooltip = ({
  children,
  trigger,
  overlay,
  mouseEnterDelay,
  mouseLeaveDelay,
  popupStyle,
  popupTransitionName,
  zIndex,
  popupAlign,
  popupClassName,
  destroyPopupOnHide,
  placement
}: Props) => {
  const TooltipBase = () => [
    <TooltipArrow className="rc-tooltip-arrow" key="arrow" />,
    <TooltipContent key="content" {...popupStyle}>
      {overlay}
    </TooltipContent>
  ];

  return (
    <Trigger
      popup={TooltipBase}
      action={trigger}
      mouseEnterDelay={mouseEnterDelay}
      mouseLeaveDelay={mouseLeaveDelay}
      popupStyle={popupStyle}
      popupTransitionName={popupTransitionName}
      zIndex={zIndex}
      popupAlign={popupAlign}
      popupClassName={popupClassName}
      destroyPopupOnHide={destroyPopupOnHide}
      popupPlacement={placement}
      builtinPlacements={placements}
    >
      {children}
    </Trigger>
  );
};

export default Tooltip;
