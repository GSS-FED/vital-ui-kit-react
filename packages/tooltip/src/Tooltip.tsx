/**
 * Copyright Â© 2018 Galaxy Software Services https://github.com/GSS-FED/vital-ui-kit-react
 * MIT license
 */

import { defaultTheme } from '@vital-ui/react-theme';
import Trigger from 'rc-trigger';
import 'rc-trigger/assets/index.css';
import * as React from 'react';
import styled, { createGlobalStyle } from 'styled-components';
import { placements } from './placements';
import { rcTooltipStyle } from './styled';

const TooltipContent = styled.div`
  display: inline-block;
  padding: 8px 16px;
  text-align: center;
  color: ${({ theme }) => theme.tooltip.color};
  background: ${({ theme }) => theme.tooltip.bg};
  border-radius: 4px;
`;

const RcTooltipStyle = createGlobalStyle`
${rcTooltipStyle};
`;

TooltipContent.defaultProps = {
  theme: defaultTheme,
};

const TooltipArrow = styled.div``;

export type Placement =
  | 'left'
  | 'right'
  | 'top'
  | 'bottom'
  | 'topLeft'
  | 'leftTop'
  | 'topRight'
  | 'rightTop'
  | 'bottomRight'
  | 'rightBottom'
  | 'bottomLeft'
  | 'leftBottom';

interface PopupAlign {
  /**
   * move point of source node to align with point of target node, such as ['tr','cc'],
   * align top right point of source node with center point of target node.
   * point can be 't'(top), 'b'(bottom), 'c'(center), 'l'(left), 'r'(right)
   */
  points: string[];
  /** offset source node by offset[0] in x and offset[1] in y. If offset contains percentage string value, it is relative to sourceNode region. */
  offset: number[];
  /** offset target node by offset[0] in x and offset[1] in y. If targetOffset contains percentage string value, it is relative to targetNode region. */
  targetOffset: number[];
  /**
   * if adjustX field is true, then will adjust source node in x direction if source node is invisible.
   * if adjustY field is true, then will adjust source node in y direction if source node is invisible.
   */
  overflow: {
    adjustX: boolean;
    adjustY: boolean;
  };
  /** whether use css right instead of left to position */
  useCssRight: boolean;
  /** whether use css bottom instead of top to position */
  useCssBottom: boolean;
  /** whether use css transform instead of left/top/right/bottom to position if browser supports. Defaults to false. */
  useCssTransform: boolean;
}

interface RcTriggerProps {
  /** whether popup is visible */
  popupVisible?: boolean;
  /** which actions cause popup shown */
  action?: Array<'hover' | 'click' | 'focus' | 'contextMenu'>;
  popupAlign?: PopupAlign;
  /** Popup content */
  popup: React.ReactNode | (() => React.ReactNode);
  /** additional style of popup */
  popupStyle?: React.CSSProperties;
  /** additional className added to popup */
  popupClassName?: string;
  /** builtin placement align map. used by placement prop */
  builtinPlacements?: object;
  zIndex?: number;
  /** https://github.com/react-component/animate */
  popupTransitionName?: string | object;
  maskTransitionName?: string | object;
  /** delay time to show when mouse enter. unit: s. */
  mouseEnterDelay?: number;
  /** delay time to hide when mouse leave. unit: s. */
  mouseLeaveDelay?: number;
  /** whether destroy popup when hide */
  destroyPopupOnHide?: boolean;
  /** 'width', 'minWidth', 'height', 'minHeight' */
  stretch?: string;
  /** use preset popup align config from builtinPlacements, can be merged by popupAlign prop */
  popupPlacement?: any;
}

export interface TooltipProps extends RcTriggerProps {
  /** the element target for tooltip */
  children: React.ReactNode;
  placement?: Placement;
}

export const Tooltip: React.SFC<TooltipProps> = ({
  children,
  popupStyle,
  popup,
  placement,
  popupPlacement,
  ...props
}) => {
  const TooltipBase = () => [
    <TooltipArrow className="rc-tooltip-arrow" key="arrow" />,
    <TooltipContent key="content" {...popupStyle}>
      {popup}
    </TooltipContent>,
  ];

  return (
    <>
      <Trigger
        popupPlacement={popupPlacement || placement}
        popup={TooltipBase}
        builtinPlacements={placements}
        {...props}
      >
        {children}
      </Trigger>
      <RcTooltipStyle />
    </>
  );
};

Tooltip.defaultProps = {
  action: ['hover'],
  mouseEnterDelay: 0,
  mouseLeaveDelay: 0,
  popupClassName: 'vital-popup',
  destroyPopupOnHide: false,
  placement: 'bottom',
};
