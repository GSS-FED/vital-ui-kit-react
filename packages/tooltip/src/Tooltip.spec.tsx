import React from 'react';
import { render } from 'react-testing-library';
import { Tooltip } from './Tooltip';
import { Button } from '@vital-ui/react-button/src';

describe('Tooltip', () => {
  it('should render with default styles', () => {
    const tooltip = render(
      <Tooltip
        placement="left"
        popup={<div>hi</div>}
        action={['click']}
      >
        <Button>Click</Button>
      </Tooltip>,
    );

    const button = tooltip.getByText('Click');
    button.click();
    expect(tooltip).toMatchSnapshot();
  });
});
