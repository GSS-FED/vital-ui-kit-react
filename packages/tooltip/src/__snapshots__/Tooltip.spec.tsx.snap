// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render with default styles 1`] = `
.c0 .rc-tooltip-arrow {
  position: absolute;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
}

.c0 .rc-trigger-popup-placement-top .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-topLeft .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-topRight .rc-tooltip-arrow {
  bottom: -5px;
  margin-left: -5px;
  border-width: 5px 5px 0;
  border-top-color: #182d54;
}

.c0 .rc-trigger-popup-placement-top .rc-tooltip-arrow {
  left: 50%;
}

.c0 .rc-trigger-popup-placement-topLeft .rc-tooltip-arrow {
  left: 15%;
}

.c0 .rc-trigger-popup-placement-topRight .rc-tooltip-arrow {
  right: 15%;
}

.c0 .rc-trigger-popup-placement-right .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-rightTop .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-rightBottom .rc-tooltip-arrow {
  left: -5px;
  margin-top: -5px;
  border-width: 5px 5px 5px 0;
  border-right-color: #182d54;
}

.c0 .rc-trigger-popup-placement-right .rc-tooltip-arrow {
  top: 50%;
}

.c0 .rc-trigger-popup-placement-rightTop .rc-tooltip-arrow {
  top: 15%;
  margin-top: 0;
}

.c0 .rc-trigger-popup-placement-rightBottom .rc-tooltip-arrow {
  bottom: 15%;
}

.c0 .rc-trigger-popup-placement-left .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-leftTop .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-leftBottom .rc-tooltip-arrow {
  right: -5px;
  margin-top: -5px;
  border-width: 5px 0 5px 5px;
  border-left-color: #182d54;
}

.c0 .rc-trigger-popup-placement-left .rc-tooltip-arrow {
  top: 50%;
}

.c0 .rc-trigger-popup-placement-leftTop .rc-tooltip-arrow {
  top: 15%;
  margin-top: 0;
}

.c0 .rc-trigger-popup-placement-leftBottom .rc-tooltip-arrow {
  bottom: 15%;
}

.c0 .rc-trigger-popup-placement-bottom .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-bottomLeft .rc-tooltip-arrow,
.c0 .rc-trigger-popup-placement-bottomRight .rc-tooltip-arrow {
  top: -5px;
  margin-left: -5px;
  border-width: 0 5px 5px;
  border-bottom-color: #182d54;
}

.c0 .rc-trigger-popup-placement-bottom .rc-tooltip-arrow {
  left: 50%;
}

.c0 .rc-trigger-popup-placement-bottomLeft .rc-tooltip-arrow {
  left: 15%;
}

.c0 .rc-trigger-popup-placement-bottomRight .rc-tooltip-arrow {
  right: 15%;
}

<button
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  hover
</button>
`;
